File: Advanced/PageAndRDL/CustomPdfExport/VB.NET/PdfRendering/Rendering/PdfContentGenerator.Pen.vb
Imports PdfSharp.Drawing
Imports GrapeCity.ActiveReports.Drawing
Imports System.Linq
Imports System.Drawing
Imports System

Partial Class PdfContentGenerator
	Private Class Pen
		Inherits PenEx
		Public Overrides Property Color() As Color
		Public Overrides Property Width() As Single
		Public Overrides Property DashStyle() As PenStyleEx
		Public Overrides Property Alignment() As PenAlignment
		Public Overrides Property StartCap() As LineCap
		Public Overrides Property EndCap() As LineCap
		Public Overrides Property DashCap() As DashCap
		Public Overrides Property LineJoin() As LineJoin
		Public Overrides Property DashPattern() As Single()

		Public Overrides Function Clone() As Object
			Return New Pen() With {
					.Color = Color,
					.Width = Width,
					.DashStyle = DashStyle,
					.Alignment = Alignment,
					.StartCap = StartCap,
					.EndCap = EndCap,
					.DashCap = DashCap,
					.LineJoin = LineJoin,
					.DashPattern = DashPattern
				}
		End Function

		Public Shared Widening Operator CType(pen As Pen) As XPen
			Dim currentPen As New XPen(XColor.FromArgb(pen.Color), pen.Width / PdfConverter.TwipsPerPoint)
			currentPen.DashStyle = CType([Enum].Parse(GetType(XDashStyle), pen.DashStyle.ToString()), XDashStyle)
			currentPen.LineCap = CType([Enum].Parse(GetType(XLineCap), pen.StartCap.ToString()), XLineCap)
			If pen.LineJoin = GrapeCity.ActiveReports.Drawing.LineJoin.MiterClipped Then
				currentPen.LineJoin = XLineJoin.Miter
			Else
				currentPen.LineJoin = CType([Enum].Parse(GetType(XLineJoin), pen.LineJoin.ToString()), XLineJoin)
			End If
			If pen.DashPattern IsNot Nothing Then
				currentPen.DashPattern = pen.DashPattern.Cast(Of Double)().ToArray()
			End If
			Return currentPen
		End Operator
	End Class
End Class

File: Advanced/PageAndRDL/CustomPdfExport/C#/PdfRendering/Rendering/PdfContentGenerator.Pen.cs
using System;
using System.Drawing;
using System.Linq;
using GrapeCity.ActiveReports.Drawing;
using PdfSharp.Drawing;

namespace GrapeCity.ActiveReports.Samples.Export.Rendering.PdfSharp
{
	partial class PdfContentGenerator
	{
		private class Pen : PenEx
		{
			public override Color Color { get; set; }
			public override float Width { get; set; }
			public override PenStyleEx DashStyle { get; set; }
			public override PenAlignment Alignment { get; set; }
			public override LineCap StartCap { get; set; }
			public override LineCap EndCap { get; set; }
			public override DashCap DashCap { get; set; }
			public override LineJoin LineJoin { get; set; }
			public override float[] DashPattern { get; set; }

			public override object Clone()
			{
				return new Pen
				{
					Color = Color,
					Width = Width,
					DashStyle = DashStyle,
					Alignment = Alignment,
					StartCap = StartCap,
					EndCap = EndCap,
					DashCap = DashCap,
					LineJoin = LineJoin,
					DashPattern = DashPattern
				};
			}

			public static implicit operator XPen(Pen pen)
			{
				var currentPen = new XPen(XColor.FromArgb(pen.Color.ToArgb()), pen.Width / PdfConverter.TwipsPerPoint);
				currentPen.DashStyle = (XDashStyle)Enum.Parse(typeof(XDashStyle), pen.DashStyle.ToString());
				currentPen.LineCap = (XLineCap)Enum.Parse(typeof(XLineCap), pen.StartCap.ToString());
				if (pen.LineJoin == LineJoin.MiterClipped)
					currentPen.LineJoin = XLineJoin.Miter;
				else
					currentPen.LineJoin = (XLineJoin)Enum.Parse(typeof(XLineJoin), pen.LineJoin.ToString());
				if (pen.DashPattern != null)
					currentPen.DashPattern = pen.DashPattern.Cast<double>().ToArray();
				return currentPen;
			}
		}
	}
}
